proposals/0021-root-signature-in-clang.md

```diff
@@ -0,0 +1,264 @@
+
+```
+
+  ; named metadata for entry root signature
+  !hlsl.entry.rootsignatures = !{!2}
+  …
+  ; link the global variable to entry function
+  !2 = !{ptr @main,
+    %0 {
+       %"dxbc::RootSignature::ContainerRootSignatureDesc" 
+          { i32 2, i32 0, i32 24, i32 0, i32 96, i32 3 }, 
+      [6 x %"dxbc::RootSignature::ContainerRootParameter"]  
+        [%"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 2, i32 0, i32 200 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 3, i32 0, i32 212 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 4, i32 0, i32 224 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 0, i32 0, i32 236 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 0, i32 0, i32 316 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 1, i32 0, i32 348 }], 
+      [2 x %"dxbc::RootSignature::StaticSamplerDesc"]
+         [%"dxbc::RootSignature::StaticSamplerDesc" 
+           { i32 85, i32 1, i32 1, i32 1, float 0.000000e+00, i32 16, i32 4, 
+             i32 2, float 0.000000e+00, float 0x47EFFFFFE0000000, i32 1, 
+             i32 0, i32 0 }, 
+          %"dxbc::RootSignature::StaticSamplerDesc" 
+           { i32 21, i32 3, i32 1, i32 1, float 0.000000e+00, i32 16, i32 4, 
+           i32 2, float 0.000000e+00, float 0x47EFFFFFE0000000, i32 2, 
+           i32 0, i32 0 }], 
+      %1 { 
+        %"dxbc::RootSignature::ContainerRootDescriptor" 
+          { i32 0, i32 1, i32 8 }, 
+        %"dxbc::RootSignature::ContainerRootDescriptor" 
+          zeroinitializer, 
+        %"dxbc::RootSignature::ContainerRootDescriptor" 
+          zeroinitializer, 
+        %2 { %"dxbc::RootSignature::ContainerRootDescriptorTable" 
+        { i32 3, i32 244 }, 
+        [3 x %"dxbc::RootSignature::ContainerDescriptorRange"] 
+         [%"dxbc::RootSignature::ContainerDescriptorRange" 
+           { i32 2, i32 1, i32 1, i32 0, i32 0, i32 -1 }, 
+          %"dxbc::RootSignature::ContainerDescriptorRange" 
+           { i32 0, i32 8, i32 1, i32 0, i32 1, i32 -1 }, 
+          %"dxbc::RootSignature::ContainerDescriptorRange" 
+           { i32 1, i32 -1, i32 1, i32 0, i32 1, i32 -1 }] 
+       }, 
+       %3 {
+           %"dxbc::RootSignature::ContainerRootDescriptorTable" 
+           { i32 1, i32 324 }, 
+           [1 x %"dxbc::RootSignature::ContainerDescriptorRange"] 
+           [%"dxbc::RootSignature::ContainerDescriptorRange" 
+             { i32 3, i32 4, i32 0, i32 1, i32 0, i32 -1 }] }, 
+           %"dxbc::RootSignature::ContainerRootConstants" 
+             { i32 10, i32 0, i32 3 } 
+       }
+      }
+    }
+
+```
+
+* LLVM IR for root signature saved in metadata.
+
+```
+
+  ; named metadata for entry root signature
+  !hlsl.entry.rootsignatures = !{!3}
+  …
+  !3 = !{ptr @main, !4}
+  !4 = !{i32 2, i32 3, !5, !18}      ; version 1.1, flag 3, root params, statis samplers.
```

Comment:
How is this version number determined?  It isn't specified in the string passed to the attribute in the HLSL.

What is it used for?  Seems that this is more a detail when the RS is serialized.

Reply:
I had a similar reaction to seeing the version, but I think you can specify the version from the frontend if you use the `rootsig_1_0` or `rootsig_1_1` target profile. So we would need it here.

Reply:
This would be good information to capture in the spec.